R/dependencies.R

```diff
@@ -554,3 +556,103 @@ check_yamls_consistent <- function(dep_structure) {
+#'
+#'
+#' @param available_packages (`data.frame`) A dataframe of the format given by
+#'   `as.data.frame(utils::available.packages)`. It is unlikely this default needs to be changed;
+#'   however you need to ensure the `options("repos")` contains the urls of all expected repos
+#'   (e.g. Bioconductor).
+#' @param include_suggests (`logical`) when considering the external packages, should the 'suggests'
+#'   packages be considered as dependencies. (This does not affect the Suggests packages of your
+#'   internal packages which are always included)
+#' @inheritParams build_check_install
+#' @md
+#' @return A vector of 'external' R packages required to install
+#'   the selected 'internal' packages. This can be used with `remotes::system_requirements`
+#'   to extract the system requirements needed for your packages, see example below.
+#' @examples
+#' \dontrun{
+#'   x <- dependency_table("openpharma/stageddeps.electricity",
+#'     project_type = "repo@@host", feature = "main")
+#'
+#'   # get external package dependencies
+#'   ex_deps <- get_all_external_dependencies(x)
+#'   print(ex_deps)
+#'
+#'   # get system dependencies (in this case there are none)
+#'   unique(unlist(lapply(ex_deps,
+#'     function(pkg, ...)
+#'       remotes::system_requirements(package = pkg, ...),
+#'     os = "ubuntu",
+#'     os_release = "20.04")
+#'  ))
+#' }
+#' @export
+get_all_external_dependencies <- function(dep_structure,
+                                          available_packages = as.data.frame(utils::available.packages()),
+                                          install_direction = "upstream",
+                                          packages_to_process = NULL,
+                                          include_suggests = FALSE) {
+  stopifnot(methods::is(dep_structure, "dependency_structure"))
+  stopifnot(is.logical(include_suggests) && length(include_suggests) == 1)
+
+  stopifnot(methods::is(available_packages, "data.frame"))
+  stopifnot(all(c("Depends", "Suggests", "Imports", "Package") %in% colnames(available_packages)))
+
+  if (!all(install_direction %in% dep_structure$direction)) {
+    stop("Invalid install_direction argument for this dependency object")
+  }
+
+  r_core_packages <- c("base", "compiler", "datasets", "graphics", "grDevices", "grid",
+  "methods", "parallel", "splines", "stats", "stats4", "tcltk", "tools", "translations", "utils")
+
+  # get the packages to consider
+  pkg_df <- dep_structure$table
+
+  pkg_names <- filter_pkgs(pkg_df, install_direction = install_direction,
+                           include_project = TRUE,
+                           dependency_packages = packages_to_process)
+
+
+  # we also need to consider the upstream dependencies of e.g. the downstream dependencies
+  # they may have been filtered out by the above
+  upstream_pkgs <- get_descendants(dep_structure$deps[["upstream_deps"]], pkg_names)
+
+  # the external packages we have considered
+  external_packages <- character(0)
```

Comment:
here you should start with all the *direct external dependencies* from `pkg_names` and `upstream_pkgs`. Also, we need two arguments

1.  `from_internal_dependencies = c("Depends", "Imports", "LinkingTo", "Suggests", "Enhances")`
2.  `from_external_dependencies = c("Depends", "Imports", "LinkingTo", "Suggests", "Enhances")`

where the internal dependencies is used here to start with (usually includes suggests) where the external dependencies is used for recursion (usually without suggest), i.e. the minimal set of packages to install and test the internal dependencies.

Reply:
note the defaults for `1` and `2` need to be discussed. I think `1` should be ` c("Depends", "Imports", "LinkingTo", "Suggests") , and `2` shout be ` c("Depends", "Imports", "LinkingTo")`

Reply:
> here you should start with all the direct external dependencies from pkg_names and upstream_pkgs.

We do:
```
# the external packages we have considered
  external_packages <- character(0)
  # and those to consider
  packages_to_consider <- unique(unlist(dep_structure$deps$external[names(dep_structure$deps$external) %in% 
c(upstream_pkgs, pkg_names)]))
```


Adding those arguments makes sense will make change

Reply:
I've added the arguments `from_internal_dependencies` and `from_external_dependencies`

Reply:
thanks